frontend/ui/widget/qrmessage.lua

```diff
@@ -0,0 +1,117 @@
+--[[--
+Widget that displays a qr code.
+
+It vanishes on key press or after a given timeout.
+
+Example:
+    local UIManager = require("ui/uimanager")
+    local _ = require("gettext")
+    local Screen = require("device").screen
+    local sample
+    sample = QRMessage:new{
+        text = _("my message"),
+        height = Screen:scaleBySize(400),
+        width = Screen:scaleBySize(400),
+        timeout = 5,  -- This widget will vanish in 5 seconds.
+    }
+    UIManager:show(sample)
+]]
+
+local Blitbuffer = require("ffi/blitbuffer")
+local CenterContainer = require("ui/widget/container/centercontainer")
+local Device = require("device")
+local FrameContainer = require("ui/widget/container/framecontainer")
+local Geom = require("ui/geometry")
+local GestureRange = require("ui/gesturerange")
+local QRWidget = require("ui/widget/qrwidget")
+local InputContainer = require("ui/widget/container/inputcontainer")
+local UIManager = require("ui/uimanager")
+local Input = Device.input
+local Screen = Device.screen
+
+local QRMessage = InputContainer:new{
+    modal = true,
+    timeout = nil, -- in seconds
+    text = nil,  -- The text to encode.
+    width = nil,  -- The width. Keep it nil to use original width.
+    height = nil,  -- The height. Keep it nil to use original height.
+    dismiss_callback = function() end,
+    alpha = nil,
+    scale_factor = 0,
+}
+
+function QRMessage:init()
+    if Device:hasKeys() then
+        self.key_events = {
+            AnyKeyPressed = { { Input.group.Any },
+                seqtext = "any key", doc = "close dialog" }
+        }
+    end
+    if Device:isTouchDevice() then
+        self.ges_events.TapClose = {
+            GestureRange:new{
+                ges = "tap",
+                range = Geom:new{
+                    x = 0, y = 0,
+                    w = Screen:getWidth(),
+                    h = Screen:getHeight(),
+                }
+            }
+        }
+    end
+
+    local image_widget = QRWidget:new{
+        text = self.text,
+        width = .9 * self.width,
```

Comment:
Will require rounding, c.f., every other instance of similar computations in the codebase ;).